README.md
![author](https://img.shields.io/badge/AUTHOR-EMRE%20UTKU%20UYGUC-red)
![author](https://img.shields.io/badge/CONTACT-emreuyguc@gmail.com-yellowgreen)
![author](https://img.shields.io/badge/-FLUTTER-blue)
![GitHub](https://img.shields.io/github/license/emreuyguc/Flutter_animated_sliced_button)

# Flutter Animated Sliced Button

New Animated Sliced Button Widget.
It is very primitive now, but I will work for asymmetrical fragmentation and dispersion in different ways as soon as possible.

## Updates
v1.2:
    -fix stack overflow parameter

v1.1 :  
    -fix setState wrong shape choose.  
    -add borderRadius  
    -change widget pieces build moved to initState method

## Screenshots

![](https://i.ibb.co/1MgQQxY/ezgif-6-a2789b51f4b6.gif)



CHANGELOG.md
## [2.0.0-nullsafety] - Release 05.02.2023
    -nullsafety

## [1.2.0] - Release 22.01.2023
    -fix stack overflow parameter

## [1.1.1] - Release 03.04.2021
    -fix hotreload text length bug

## [1.1.0] - Release 03.04.2021
	-fix setState wrong shape choose.
    -add borderRadius
    -change widget pieces build moved to initState method

## [1.0.1] - Release 17.09.2020
	-add readme



lib/piece.dart
import 'package:animated_sliced_button/piece_types.dart';
import 'package:clippy_flutter/clippy_flutter.dart';
import 'package:flutter/material.dart';

class Piece extends StatelessWidget {
  final Color? backgroundColor;
  final String? char;
  final TextStyle? charStyle;
  final double? width;
  final double? height;
  final double? clipHeight;
  final dynamic shape;
  late Widget renderWidget;
  final int? index;
  final Radius? borderRadius;
  Piece(
      {this.shape,
      this.width,
      this.height,
      this.char,
      this.charStyle,
      this.backgroundColor,
      this.clipHeight,
      this.index, this.borderRadius,
      }) {
    final Widget shapeContainer = Container(
      decoration: BoxDecoration(
        color: this.backgroundColor,
      ),
      alignment: Alignment.center,
      width: this.width,
      height: this.height,
      child: Text(
        this.char!,
        style: this.charStyle,
      ),
    );

    if (this.shape.runtimeType == DIAGONAL) {
      this.renderWidget = ClipRRect(
        borderRadius: BorderRadius.only(
            topLeft: this.shape == DIAGONAL.TOP_RIGHT || this.shape == DIAGONAL.BOTTOM_RIGHT ? this.borderRadius ?? Radius.circular(0) : Radius.circular(0),
            bottomLeft: this.shape == DIAGONAL.TOP_RIGHT || this.shape == DIAGONAL.BOTTOM_RIGHT ? this.borderRadius ?? Radius.circular(0) : Radius.circular(0),
            bottomRight: this.shape == DIAGONAL.TOP_LEFT || this.shape == DIAGONAL.BOTTOM_LEFT ? this.borderRadius ?? Radius.circular(0):Radius.circular(0),
            topRight: this.shape == DIAGONAL.TOP_LEFT || this.shape == DIAGONAL.BOTTOM_LEFT ? this.borderRadius ?? Radius.circular(0):Radius.circular(0)),
        child: Diagonal(
          axis: Axis.vertical,
          clipHeight: this.clipHeight!,
          position: (this.shape as DIAGONAL).EDGE!,
          child: shapeContainer,
        ),
      );
    } else {
      SHAPE_WIDGET _shapeBody = (this.shape as SHAPE_BODY).WIDGET!;
      if (_shapeBody == SHAPE_WIDGET.PARALLEL) {
        this.renderWidget = Parallelogram(
          cutLength: this.clipHeight!,
          edge: (this.shape as SHAPE_BODY).EDGE!,
          child: shapeContainer,
        );
      }
      if (_shapeBody == SHAPE_WIDGET.TRIPAZOID) {
        this.renderWidget = Trapezoid(
          cutLength: this.clipHeight!,
          edge: (this.shape as SHAPE_BODY).EDGE!,
          child: shapeContainer,
        );
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return this.renderWidget;
  }
}


lib/piece_types.dart
import 'package:clippy_flutter/clippy_flutter.dart';

enum PIECE_TYPE { START, END, BODY }

enum SHAPE_WIDGET { DIAGONAL, TRIPAZOID, PARALLEL }

enum DIAGONAL { TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT }


enum SHAPE_BODY {
  TRIPAZOID_TOP,
  TRIPAZOID_BOTTOM,
  PARALLEL_LEFT,
  PARALLEL_RIGHT
}

extension PIECE_OPS on PIECE_TYPE {
  List<DIAGONAL>? get SHAPES {
    switch (this) {
      case PIECE_TYPE.START:
        return [DIAGONAL.BOTTOM_RIGHT, DIAGONAL.TOP_RIGHT];

      case PIECE_TYPE.END:
        return [DIAGONAL.BOTTOM_LEFT, DIAGONAL.TOP_LEFT];
    }
  }
}

extension DIAGONAL_OPS on DIAGONAL {
  List<SHAPE_BODY>? get NEXT {
    switch (this) {
      case DIAGONAL.TOP_RIGHT:
        return [SHAPE_BODY.TRIPAZOID_TOP, SHAPE_BODY.PARALLEL_RIGHT];
      case DIAGONAL.BOTTOM_RIGHT:
        return [SHAPE_BODY.TRIPAZOID_BOTTOM, SHAPE_BODY.PARALLEL_LEFT];
    }
  }

  DiagonalPosition? get EDGE {
    switch (this) {
      case DIAGONAL.TOP_RIGHT:
        return DiagonalPosition.TOP_RIGHT;
      case DIAGONAL.BOTTOM_RIGHT:
        return DiagonalPosition.BOTTOM_RIGHT;
      case DIAGONAL.TOP_LEFT:
        return DiagonalPosition.TOP_LEFT;
      case DIAGONAL.BOTTOM_LEFT:
        return DiagonalPosition.BOTTOM_LEFT;
    }
  }

  get WIDGET {
    return SHAPE_WIDGET.DIAGONAL;
  }
}

extension SHAPE_BODY_OPS on SHAPE_BODY {
  List<SHAPE_BODY>? get NEXT {
    switch (this) {
      case SHAPE_BODY.TRIPAZOID_TOP:
        return [SHAPE_BODY.PARALLEL_LEFT, SHAPE_BODY.TRIPAZOID_BOTTOM];

      case SHAPE_BODY.TRIPAZOID_BOTTOM:
        return [SHAPE_BODY.PARALLEL_RIGHT, SHAPE_BODY.TRIPAZOID_TOP];

      case SHAPE_BODY.PARALLEL_LEFT:
        return [SHAPE_BODY.TRIPAZOID_BOTTOM];

      case SHAPE_BODY.PARALLEL_RIGHT:
        return [SHAPE_BODY.TRIPAZOID_TOP];
    }
  }

  DIAGONAL? get END {
    switch (this) {
      case SHAPE_BODY.TRIPAZOID_TOP:
      case SHAPE_BODY.PARALLEL_LEFT:
        return DIAGONAL.TOP_LEFT;

      case SHAPE_BODY.TRIPAZOID_BOTTOM:
      case SHAPE_BODY.PARALLEL_RIGHT:
        return DIAGONAL.BOTTOM_LEFT;
    }
  }

  Edge? get EDGE {
    switch (this) {
      case SHAPE_BODY.TRIPAZOID_TOP:
        return Edge.TOP;

      case SHAPE_BODY.TRIPAZOID_BOTTOM:
        return Edge.BOTTOM;

      case SHAPE_BODY.PARALLEL_LEFT:
        return Edge.LEFT;

      case SHAPE_BODY.PARALLEL_RIGHT:
        return Edge.RIGHT;
    }
  }

  SHAPE_WIDGET? get WIDGET {
    switch (this) {
      case SHAPE_BODY.TRIPAZOID_BOTTOM:
      case SHAPE_BODY.TRIPAZOID_TOP:
        return SHAPE_WIDGET.TRIPAZOID;

      case SHAPE_BODY.PARALLEL_RIGHT:
      case SHAPE_BODY.PARALLEL_LEFT:
        return SHAPE_WIDGET.PARALLEL;
    }
  }
}


example/lib/main.dart
import 'package:animated_sliced_button/animated_sliced_button.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Crash Button Demo',
      home: MyHomePage(title: 'Crash Widget Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        width: double.infinity,
        height: double.infinity,
        color: Colors.grey,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Container(
              alignment: Alignment.center,
              color: Colors.black,
              width: 300,
              height: 100,
              child: Text(
                "E.U.U",
                style: TextStyle(color: Colors.amber, fontSize: 40),
              ),
            ),
            SizedBox(
              height: 20,
            ),
            SlicedButton(
              borderRadius: Radius.circular(10),
              width: 300,
              height: 100,
              crashDuration: Duration(milliseconds: 150),
              crashDistance: 20,
              textStyle: TextStyle(color: Colors.cyanAccent, fontSize: 30),
              text: "SLICED",
              backgroundColor: Colors.black87,
              onTap: () {
                print("bum");
              },
            ),
            SizedBox(
              height: 20,
            ),
            SlicedButton(
              borderRadius: Radius.circular(10),
              width: 300,
              height: 100,
              crashDuration: Duration(milliseconds: 150),
              crashDistance: 20,
              textStyle: TextStyle(color: Colors.cyanAccent, fontSize: 30),
              text: "WIDGET",
              backgroundColor: Colors.black87,
              onTap: () {
                print("bum");
              },
            ),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/animated_sliced_button.dart
library animated_sliced_button;

import 'package:animated_sliced_button/piece.dart';
import 'package:animated_sliced_button/piece_types.dart';
import 'package:flutter/material.dart';

class SlicedButton extends StatefulWidget {
  final String text;
  final TextStyle? textStyle;
  final Color backgroundColor;
  final double width;
  final double height;
  final double crashDistance;
  final Duration? crashDuration;
  final VoidCallback onTap;
  final VoidCallback? onEnd;
  final Radius? borderRadius;

  SlicedButton(
      {required this.width,
      required this.height,
      required this.text,
      this.backgroundColor = Colors.black87,
      this.crashDistance = 14,
      this.crashDuration,
      this.textStyle,
      required this.onTap,
      this.onEnd,
      this.borderRadius});

  @override
  _SlicedButtonState createState() => _SlicedButtonState();
}

class _SlicedButtonState extends State<SlicedButton> {
  bool animate = false;
  late List<dynamic> _buttonShapes;

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _buildButtonShapes();
  }

  void _buildButtonShapes() {
    SHAPE_BODY? _choosedNext;
    DIAGONAL? _starter;
    dynamic _choosed;

    _buttonShapes = List<dynamic>.generate(widget.text.length, (charIndex) {
      if (charIndex == 0) {
        _starter = (PIECE_TYPE.START.SHAPES!..shuffle()).first;
        _choosed = _starter;
      } else if (charIndex == 1) {
        _choosedNext = (_starter!.NEXT!..shuffle()).first;
        _choosed = _choosedNext;
      } else if (charIndex == widget.text.length - 1) {
        DIAGONAL? _end = _choosedNext!.END;
        _choosed = _end;
      } else {
        List<SHAPE_BODY>? _body = _choosedNext!.NEXT;
        _choosedNext = (_body!..shuffle()).first;
        _choosed = _choosedNext;
      }

      return _choosed;
    });
  }

  @override
  Widget build(BuildContext context) {
    if (_buttonShapes.length != widget.text.length) {
      _buildButtonShapes();
    }

    return GestureDetector(
        onTap: () {
          setState(() {
            this.animate = !this.animate;
          });
          widget.onTap();
        },
        child: AnimatedContainer(
          duration: widget.crashDuration == null
              ? Duration(milliseconds: 150)
              : widget.crashDuration!,
          curve: Curves.easeOutBack,
          height: widget.height,
          width: widget.width,
          transform: Matrix4.translationValues(
              this.animate == true
                  ? -((widget.crashDistance * (widget.text.length - 1)) / 2)
                  : 0,
              0,
              0),
          child: Stack(
            clipBehavior: Clip.none,
            children: List<Widget>.generate(this._buttonShapes.length, (index) {
              return AnimatedPositioned(
                onEnd: () {
                  setState(() {
                    this.animate = false;
                  });
                },
                duration: Duration(milliseconds: 150),
                curve: Curves.easeOutBack,
                left: this.animate == true
                    ? ((widget.width +
                                (widget.crashDistance *
                                    (widget.text.length - 1))) /
                            widget.text.length) *
                        index
                    : (((widget.width / widget.text.length) -
                                widget.crashDistance / widget.text.length) *
                            index) *
                        0.995,
                top: this.animate == true
                    ? (index % 2 == 0
                        ? widget.crashDistance
                        : -widget.crashDistance)
                    : 0,
                //transform: Matrix4.translationValues(this.animate == true ? -(widget.crashDistance*widget.text.length*0.35) : -(index * widget.crashDistance), this.animate == true ? (index % 2 == 0 ? 20 : -20) : 0, 0),
                child: Piece(
                    borderRadius: widget.borderRadius,
                    index: index,
                    height: widget.height,
                    width: ((widget.width +
                            (widget.crashDistance * (widget.text.length - 1))) /
                        widget.text.length),
                    clipHeight: widget.crashDistance,
                    shape: _buttonShapes[index],
                    char: widget.text[index],
                    backgroundColor: widget.backgroundColor,
                    charStyle: widget.textStyle == null
                        ? TextStyle(color: Colors.cyanAccent, fontSize: 30)
                        : widget.textStyle),
              );
            }),
          ),
        ));
  }
}
